File: 0001-0050/0016/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int ans = nums[0] + nums[1] + nums[2];
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size(); i++) {
            int l = i + 1, r = nums.size() - 1;
            while (l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                if (sum == target) return sum;
                if (abs(sum - target) < abs(ans - target)) ans = sum;
                if (sum < target) l++;
                else r--;
            }
        }
        return ans;
    }
};

struct Example {
    vector<int> nums;
    int target;
    int ans;
};

int main() {
    vector<Example> examples = {
        {{-1, 2, 1, -4}, 1, 2},
        {{0, 0, 0}, 1, 0},
    };
    
    Solution *solve = new Solution();
    for (int i = 0; i < examples.size(); i++) {
        int ans = solve->threeSumClosest(examples[i].nums, examples[i].target);
        if (ans == examples[i].ans) {
            cout << "PASS: CASE " << i << endl;
        } else {
            cout << "FAIL: CASE " << i << endl;
        }
    }
    
    return 0;
}

File: 0001-0050/0016/solution.py
from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        ans = nums[0] + nums[1] + nums[2]
        nums.sort()
        for i in range(len(nums)):
            l, r = i + 1, len(nums) - 1
            while l < r:
                _sum = nums[i] + nums[l] + nums[r]
                if _sum == target:
                    return _sum
                if abs(_sum - target) < abs(ans - target):
                    ans = _sum
                if _sum < target:
                    l += 1
                else:
                    r -= 1
        return ans


class Example:
    def __init__(self, nums, target, ans):
        self.nums = nums
        self.target = target
        self.ans = ans


def main():
    examples = [
        Example([-1, 2, 1, -4], 1, 2),
        Example([0, 0, 0], 1, 0)
    ]

    solve = Solution()
    for i, example in enumerate(examples):
        ans = solve.threeSumClosest(example.nums, example.target)
        if ans == example.ans:
            print(f"PASS: CASE {i}")
        else:
            print(f"FAIL: CASE {i}")


if __name__ == "__main__":
    main()
